MoltenVK/MoltenVK/Commands/MVKCmdQueries.mm

```diff
@@ -117,10 +117,13 @@
 }
 
 void MVKCmdCopyQueryPoolResults::encode(MVKCommandEncoder* cmdEncoder) {
+    MVKCountingEvent* pCmdBuffDoneEvent = cmdEncoder->_cmdBuffDoneEvent.get();
+    pCmdBuffDoneEvent->lock();
     [cmdEncoder->_mtlCmdBuffer addCompletedHandler: ^(id<MTLCommandBuffer> mtlCmdBuff) {
         // This can block, so it must not run on the Metal completion queue.
-        dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
+        dispatch_async(dispatch_get_global_queue(QOS_CLASS_USER_INITIATED, 0), ^{
             _queryPool->copyQueryPoolResults(_query, _queryCount, _destBuffer, _destOffset, _destStride, _flags);
+            pCmdBuffDoneEvent->unlock();
         });
     }];
 }
```

Comment:
> This PR is making me rethink how queries are handled in MoltenVK. We only effectively support occlusion queries. Timestamps are pretty lamely supported. And the use of CPU code for this command seems to be a catalyst for much of the trouble you are dealing with right now.
> 
> How much would it help if we moved this copy command operation to a BLIT or compute shader?
> 
> I'm wondering that if we moved this to a BLIT or compute shader, that it would remove the race condition you are encountering between this copy command and `finishQueries()`.
> 


That would probably help immensely. Frankly, I'm wondering why it wasn't done this way to begin with.

> We'd still need the existing `MVKQueryPool::getResults()` code for the `vkGetQueryPoolResults()` function...but it doesn't need to deal with the challenges of blocking during command buffer execution. And `finishQueries()` would only be used by `vkGetQueryPoolResults()`, so we might be able to drastically simplify the blocking of that function for query availability.
> 
> By using a compute shader, we might have to give up the ability to block for query availability on this command...since it would be a GPU block...but for occlusion queries, the render pass will be done by the time this copy command executes and I'm wondering if we could just blindly assume all occlusion queries were "available".

I think that might work.

> 
> For timestamps...we could either abandon support for them altogether...or mark them at the point the command is submitted to the encoder...and again...assume that by the time this copy command runs all the timestamp queries are "available".

Command buffers have `GPUStartTime` and `GPUEndTime` properties, but these are only available on iOS.

Reply:
>That would probably help immensely. 

Do you want to take care of that as part of a rework to this PR, or do you want me to add that first on a separate PR, so you can leverage it for a rework of this PR?

>Frankly, I'm wondering why it wasn't done this way to begin with.

Just history. From the early interpretations of Vulkan synchronization as it applies to Metal.

>Command buffers have GPUStartTime and GPUEndTime properties, but these are only available on iOS.

Yeah...I looked at those too. Aside from the platform issue...they only seem to be available during the completion callback. They seem more suited as a means of tracking command buffer execution duration after the fact...as opposed to time stamping. It didn't seem to be any better than marking timestamps during the completion callback itself.

Reply:
> Do you want to take care of that as part of a rework to this PR, or do you want me to add that first on a separate PR, so you can leverage it for a rework of this PR?

I'll work on adding that to this PR.